packages/typescript-group-checker/src/compilers/compiler-with-watch.ts

```diff
@@ -0,0 +1,154 @@
+    this.allTSConfigFiles = new Set([path.resolve(this.tsconfigFile)]);
+  }
+
+  public async init(): Promise<{ dependencyFiles: DependencyFile[]; errors: ts.Diagnostic[] }> {
+    guardTSVersion();
+    this.guardTSConfigFileExists();
+    const buildModeEnabled = determineBuildModeEnabled(this.tsconfigFile);
+
+    const host = ts.createSolutionBuilderWithWatchHost(
+      {
+        ...ts.sys,
+        readFile: (fileName) => {
+          const content = this.fs.getFile(fileName)?.content;
+          if (content && this.allTSConfigFiles.has(path.resolve(fileName))) {
+            return this.adjustTSConfigFile(fileName, content, buildModeEnabled);
+          }
+          return content;
+        },
+        getModifiedTime: (pathName: string) => {
+          return this.fs.getFile(pathName)?.modifiedTime;
+        },
+        watchFile: (filePath: string, callback: ts.FileWatcherCallback) => {
+          const file = this.fs.getFile(filePath);
+          if (file) {
+            file.watcher = callback;
+          }
+
+          return {
+            close: () => {
+              delete this.fs.getFile(filePath)!.watcher;
+            },
+          };
+        },
+        writeFile: (fileName, data) => {
+          this.fs.writeFile(fileName, data);
+        },
+      },
+      undefined,
+      (error) => {
+        this.currentErrors.push(error);
+      },
+      (status) => {
+        this.log.debug(status.messageText.toString());
+      },
+      (summary) => {
+        this.currentTask.resolve();
+      }
+    );
+
+    host.afterProgramEmitAndDiagnostics = (program) => {
+      const sourceFiles = program.getSourceFiles().map((f: any) => {
+        const file: { fileName: string; imports: Array<{ text: string }> } = f;
+
+        const result = {
+          fileName: this.resolveFilename(file.fileName),
+          imports: new Set(
+            file.imports
+              .filter((i) => {
+                if (i.text) {
+                  return true;
+                } else {
+                  return false;
+                }
+              })
+              .map((i) => ts.resolveModuleName(i.text, file.fileName, {}, host).resolvedModule?.resolvedFileName ?? `${i.text}-not_resolved`)
+              .map((i) => this.resolveFilename(i))
+          ),
+        };
+
+        return result;
+      });
+
+      for (const sourceFile of sourceFiles) {
+        const indexInSourceFileArray = this.sourceFiles.findIndex((sourceFileThis) => sourceFileThis.fileName === sourceFile.fileName);
+        if (indexInSourceFileArray >= 0) {
+          sourceFile.imports.forEach((sourceFileImport) => this.sourceFiles[indexInSourceFileArray].imports.add(sourceFileImport));
+        } else {
+          this.sourceFiles.push(sourceFile);
+        }
+      }
+    };
+
+    const compiler = ts.createSolutionBuilderWithWatch(host, [this.tsconfigFile], {});
+
+    compiler.build();
+
+    const errors = await this.check();
+
+    if (!this.sourceFiles) {
```

Comment:
Moeten deze niet al geset worden voordat je gaat checken? (Op regel 103 worden ze namelijk gepusht)

PS: een lege array is truthy dus als de waarde van `this.sourceFiles` gelijk is aan `[]` dan gaat deze error niet af. Is dat de bedoeling?

Reply:
Nadat typescript een keer gebuild heeft willen wij bekijken of Typescript de source files wel gevuld heeft door middel van de callback functie op regel 103. Nadat er 1 keer gecheckt is weten wij zeker dat de files gezet moeten zijn en als dat door een reden niet is dan throwen wij een error.

Het checken op een lege array was inderdaad fout 😅